src/main/java/picard/illumina/ClusterDataToSamConverter.java

```diff
@@ -206,13 +237,24 @@ private SAMRecord createSamRecord(final ReadData readData, final String readName
         final String readName = readNameEncoder.generateReadName(cluster, null); // Use null here to prevent /1 or /2 suffixes on read name.
 
         // Get and transform the unmatched barcode, if any, to store with the reads
-        String unmatchedBarcode = null;
-        if (hasSampleBarcode && cluster.getMatchedBarcode() == null) {
-            final byte[][] barcode = new byte[sampleBarcodeIndices.length][];
-            for (int i = 0; i < sampleBarcodeIndices.length; i++) {
-                barcode[i] = cluster.getRead(sampleBarcodeIndices[i]).getBases();
+        final String unmatchedBarcode;
+        final String barcodeQuality;
+        if (hasSampleBarcode) {
+            if (this.barcodePopulationStrategy == PopulateBarcode.ALWAYS ||
+                    this.barcodePopulationStrategy == PopulateBarcode.ORPHANS_ONLY && cluster.getMatchedBarcode() == null ||
+                    this.barcodePopulationStrategy == PopulateBarcode.INEXACT_MATCH && !getUnmatchedBarcode(cluster, false).equals(cluster.getMatchedBarcode())) {
+                unmatchedBarcode = getUnmatchedBarcode(cluster, true);
+            } else {
+                unmatchedBarcode = null;
             }
-            unmatchedBarcode = IlluminaUtil.barcodeSeqsToString(barcode).replace('.', 'N'); //TODO: This has a separator, where as in other places we do not use a separator
+        } else {
+            unmatchedBarcode = null;
+        }
+
+        if (unmatchedBarcode != null && includeQualitiesWithBarcode){
+            barcodeQuality = getBarcodeQuality(cluster);
+        } else {
+            barcodeQuality = null;
         }
 
         final List<String> molecularIndexes;
```

Comment:
Not in scope of this review but this is weird too. Collections are not really final so there is no reason not to just use

```       
 final List<String> molecularIndexes = new ArrayList<>();
        final List<String> molecularIndexQualities = new ArrayList<>();
        if (hasMolecularBarcode) {
            for (int i = 0; i < molecularBarcodeIndices.length; i++) {
                molecularIndexes.add(new String(cluster.getRead(molecularBarcodeIndices[i]).getBases()).replace('.', 'N'));
                molecularIndexQualities.add(SAMUtils.phredToFastq(cluster.getRead(molecularBarcodeIndices[i]).getQualities()));
            }
        }
```

Reply:
done